From 81fbb11bb06936b8a0aca007f226d62ae7715ae2 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Tue, 3 Jun 2025 12:14:07 +0100
Subject: [PATCH 7/7] fips: add back attempt to turn on FIPS mode

microsoft/go has stopped attempting to turn on FIPS mode
selectively. This breaks running fips CNI binaries on fips-capable
hosts, but with fips not turned on system wide. This used to work with
1.23 based toolchain, thus this is seen as a regression.
---
 src/crypto/internal/backend/openssl_linux.go | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/src/crypto/internal/backend/openssl_linux.go b/src/crypto/internal/backend/openssl_linux.go
index 02c3c67f82..da808f3a2b 100644
--- a/src/crypto/internal/backend/openssl_linux.go
+++ b/src/crypto/internal/backend/openssl_linux.go
@@ -25,13 +25,22 @@ const Enabled = true
 
 type BigInt = openssl.BigInt
 
+// Chainguard: we always try to enable fips, so try to make it so
+func chainguardFIPSCapable() bool {
+	if !openssl.FIPSCapable() {
+		// SetFIPS returns an error, so if we get nil it means that everything was OK.
+		return openssl.SetFIPS(true) == nil
+	}
+	return true
+}
+
 func init() {
 	// Some distributions, e.g. Azure Linux 3, don't set the `fips=yes` property when running in FIPS mode,
 	// but they configure OpenSSL to use a FIPS-compliant provider (in the case of Azure Linux 3, the SCOSSL provider).
 	// In this cases, openssl.FIPS would return `false` and openssl.FIPSCapable would return `true`.
 	// We don't care about the `fips=yes` property as long as the provider is FIPS-compliant, so use
 	// openssl.FIPSCapable to determine whether FIPS mode is enabled.
-	if err := fips140.Check(func() bool { return openssl.FIPSCapable() }); err != nil {
+	if err := fips140.Check(func() bool { return chainguardFIPSCapable() }); err != nil {
 		// This path can be reached for the following reasons:
 		// - In OpenSSL 1, the active engine doesn't support FIPS mode.
 		// - In OpenSSL 1, the active engine supports FIPS mode, but it is not enabled.
-- 
2.47.2

